src/Workspaces/Core/Portable/Remote/RemoteHostClientExtensions.cs

```diff
@@ -1,36 +1,100 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
-using System;
 using System.Threading;
 using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Execution;
+using Microsoft.CodeAnalysis.Internal.Log;
 
 namespace Microsoft.CodeAnalysis.Remote
 {
     internal static class RemoteHostClientExtensions
     {
-        [Obsolete]
-        public static Task<RemoteHostClient.Session> CreateCodeAnalysisServiceSessionAsync(
-            this RemoteHostClient client, Solution solution, CancellationToken cancellationToken)
+        /// <summary>
+        /// Create <see cref="RemoteHostClient.Session"/> for the <paramref name="serviceName"/> if possible.
+        /// otherwise, return null.
+        /// 
+        /// Creating session could fail if remote host is not available. one of example will be user killing
+        /// remote host.
+        /// </summary>
+        public static Task<RemoteHostClient.Session> TryCreateServiceSessionAsync(this RemoteHostClient client, string serviceName, CancellationToken cancellationToken)
         {
-            return CreateCodeAnalysisServiceSessionAsync(
-                client, solution, callbackTarget: null, cancellationToken: cancellationToken);
+            return client.TryCreateSessionAsync(serviceName, callbackTarget: null, cancellationToken: cancellationToken);
         }
 
-        [Obsolete]
-        public static Task<RemoteHostClient.Session> CreateCodeAnalysisServiceSessionAsync(
-            this RemoteHostClient client, Solution solution, object callbackTarget, CancellationToken cancellationToken)
+        /// <summary>
+        /// Create <see cref="SessionWithSolution"/> for the <paramref name="serviceName"/> if possible.
+        /// otherwise, return null.
+        /// 
+        /// Creating session could fail if remote host is not available. one of example will be user killing
+        /// remote host.
+        /// </summary>
+        public static Task<SessionWithSolution> TryCreateServiceSessionAsync(this RemoteHostClient client, string serviceName, Solution solution, CancellationToken cancellationToken)
```

Comment:
because whole point of what I am doing is getting rid of solution from session. basically I want session to be purely session between client and remote host. no cancellation or solution attached to it. so that I can simply pool session.

but, you wanted a way to tie all those to the session. so that is why there is this SessionWithSolution for some one who want to tie all those.

with @sharwell 's work on removed cancellation from session, this which remove solution from session and then again @sharwell 's work to pool session, we will get to the point where getting new session is really cheap most of time and caller doesn't need to care about life time of session perf except ones like symbol searcher which require session to keep alive for states.